xmppserver/src/main/java/org/jivesoftware/util/SAXReaderUtil.java

```diff
@@ -0,0 +1,287 @@
+     *
+     * Equivalent to calling <tt>readDocument(input).getRootElement()</tt>
+     *
+     * @param reader The data to be parsed as an XML Document.
+     * @return an XML Document root element.
+     */
+    public static Element readRootElement(@Nonnull final Reader reader) throws ExecutionException, InterruptedException {
+        return readDocument(reader).getRootElement();
+    }
+
+    /**
+     * Parses an XML Document from the content provided by a file.
+     *
+     * The XML parsing is delegated to a limited set of parsers. When a parser is not immediately available, this method
+     * will block until one becomes available. The availability of parsers can be optimized by using the SystemProperties
+     * in this class.
+     *
+     * @param file The data to be parsed as an XML Document.
+     * @return an XML Document.
+     */
+    public static Document readDocument(@Nonnull final File file) throws ExecutionException, InterruptedException {
+        return parserService.submit(new ParserTask(file)).get();
+    }
+
+    /**
+     * Parses an XML Document from the content provided by a file and returns the root element of the document.
+     *
+     * Equivalent to calling <tt>readDocument(input).getRootElement()</tt>
+     *
+     * @param file The data to be parsed as an XML Document.
+     * @return an XML Document root element.
+     */
+    public static Element readRootElement(@Nonnull final File file) throws ExecutionException, InterruptedException {
+        return readDocument(file).getRootElement();
+    }
+
+    /**
+     * Constructs a new SAXReader instance, configured for XMPP parsing.
+     *
+     * @return A new SAXReader instance.
+     */
+    private static SAXReader constructNewReader() throws SAXException
+    {
+        final SAXReader saxReader = new SAXReader();
+        saxReader.setEncoding( "UTF-8" );
+        saxReader.setIgnoreComments(true);
+
+        // Guard against various known vulnerabilities (see OF-2094).
+        saxReader.setEntityResolver((publicId, systemId) -> {
+            throw new IOException("External entity denied: " + publicId + " // " + systemId);
+        });
+        saxReader.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
+        saxReader.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+        saxReader.setFeature("http://xml.org/sax/features/external-general-entities", false);
+        saxReader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
+
+        return saxReader;
+    }
+
+    private static class ParserTask implements Callable<Document>
+    {
+        private static final ThreadLocal<SAXReader> localSAXReader = ThreadLocal.withInitial(()-> {
+            try {
+                return constructNewReader();
+            } catch (SAXException e) {
+                Log.error("Unable to parse XML data.", e);
```

Comment:
Not sure the log message matches reality - you're not parsing XML at this stage.